apps/ogr2ogr_lib.cpp

```diff
@@ -586,54 +586,42 @@ static std::unique_ptr<OGRGeometry> LoadGeometry(const std::string &osDS,
     if (!osWhere.empty())
         poLyr->SetAttributeFilter(osWhere.c_str());
 
-    std::unique_ptr<OGRMultiPolygon> poMP;
+    OGRGeometryCollection oGC;
+
+    const auto poSRSSrc = poLyr->GetSpatialRef();
+    if (poSRSSrc)
+    {
+        auto poSRSClone = poSRSSrc->Clone();
+        oGC.assignSpatialReference(poSRSClone);
+        poSRSClone->Release();
+    }
+
     for (auto &poFeat : poLyr)
     {
-        OGRGeometry *poSrcGeom = poFeat->GetGeometryRef();
+        auto poSrcGeom = std::unique_ptr<OGRGeometry>(poFeat->StealGeometry());
         if (poSrcGeom)
         {
-            const OGRwkbGeometryType eType =
-                wkbFlatten(poSrcGeom->getGeometryType());
-
-            if (poMP == nullptr)
+            if (!poSrcGeom->IsValid())
             {
-                poMP = cpl::make_unique<OGRMultiPolygon>();
-                const auto poSRSSrc = poSrcGeom->getSpatialReference();
-                if (poSRSSrc)
-                {
-                    auto poSRSClone = poSRSSrc->Clone();
-                    poMP->assignSpatialReference(poSRSClone);
-                    poSRSClone->Release();
-                }
-            }
-
-            if (eType == wkbPolygon)
-                poMP->addGeometry(poSrcGeom);
-            else if (eType == wkbMultiPolygon)
-            {
-                OGRMultiPolygon *poSrcMP = poSrcGeom->toMultiPolygon();
-                const int nGeomCount = poSrcMP->getNumGeometries();
-
-                for (int iGeom = 0; iGeom < nGeomCount; iGeom++)
-                {
-                    poMP->addGeometry(poSrcMP->getGeometryRef(iGeom));
-                }
+                auto poValid =
```

Comment:
Checking validity / attempting to autocorrect errors is a new behavior here, right? If so, it seems a good idea to add a test case.

Reply:
Test added